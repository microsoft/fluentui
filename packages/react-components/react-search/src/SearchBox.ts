export * from './components/SearchBox/index';
