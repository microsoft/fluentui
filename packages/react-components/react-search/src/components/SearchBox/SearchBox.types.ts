import * as React from 'react';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import type { InputOnChangeData, InputProps, InputSlots, InputState } from '@fluentui/react-input';

export type SearchBoxSlots = InputSlots & {
  /** Last element in the input, within the input border */
  dismiss?: Slot<'span'>;
};

/**
 * SearchBox Props
 */
export type SearchBoxProps = Omit<
  ComponentProps<Partial<SearchBoxSlots>, 'input'>,
  // `children` is unsupported. The rest of these native props have customized definitions.
  'children' | 'defaultValue' | 'onChange' | 'size' | 'type' | 'value'
> &
  Omit<InputProps, 'onChange'> & {
    /**
     * Custom onChange callback.
     * Will be traditionally supplied with a React.ChangeEvent<HTMLInputElement> for usual character entry.
     * When the dismiss button is clicked, this will be called with an event of type React.MouseEvent<HTMLSpanElement>
     * and an empty string as the `value` property of the data parameter
     */
    // eslint-disable-next-line @nx/workspace-consistent-callback-type -- can't change type of existing callback
    onChange?: (event: SearchBoxChangeEvent, data: InputOnChangeData) => void;
  };

/**
 * State used in rendering SearchBox
 */
export type SearchBoxState = ComponentState<SearchBoxSlots> &
  InputState &
  Required<Pick<InputState, 'size'>> &
  Required<Pick<SearchBoxProps, 'disabled'>> & {
    focused: boolean;
  };

/** Overloaded onChange event type, used to merge functionality of regular text entry and the dismiss button */
export type SearchBoxChangeEvent = React.ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLSpanElement>;
