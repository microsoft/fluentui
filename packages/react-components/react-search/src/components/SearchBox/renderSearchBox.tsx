/** @jsxRuntime classic */
/** @jsx createElement */
/** @jsxFrag React.Fragment */

import * as React from 'react';
import { createElement } from '@fluentui/react-jsx-runtime';
import { getSlotsNext } from '@fluentui/react-utilities';
import type { SearchBoxState, SearchBoxSlots } from './SearchBox.types';

/**
 * Render the final JSX of SearchBox
 */
export const renderSearchBox_unstable = (state: SearchBoxState) => {
  const { slots, slotProps } = getSlotsNext<SearchBoxSlots>(state);

  // TODO Add additional slots in the appropriate place
  const rootSlots = {
    contentAfter: slots.contentAfter && {
      ...slotProps.contentAfter,
      children: (
        <>
          {slotProps.contentAfter.children}
          {slots.dismiss && <slots.dismiss {...slotProps.dismiss} />}
        </>
      ),
    },
  };

  return <slots.root {...slotProps.root} {...rootSlots} />;
};
