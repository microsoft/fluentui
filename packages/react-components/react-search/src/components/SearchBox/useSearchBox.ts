import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { SearchBoxProps, SearchBoxState } from './SearchBox.types';

/**
 * Create the state required to render SearchBox.
 *
 * The returned state can be modified with hooks such as useSearchBoxStyles_unstable,
 * before being passed to renderSearchBox_unstable.
 *
 * @param props - props from this instance of SearchBox
 * @param ref - reference to root HTMLElement of SearchBox
 */
export const useSearchBox_unstable = (props: SearchBoxProps, ref: React.Ref<HTMLElement>): SearchBoxState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: getNativeElementProps('div', {
      ref,
      ...props,
    }),
  };
};
