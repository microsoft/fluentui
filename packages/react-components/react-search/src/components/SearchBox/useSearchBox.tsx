import * as React from 'react';
import {
  isResolvedShorthand,
  mergeCallbacks,
  slot,
  useControllableState,
  useEventCallback,
  useMergedRefs,
} from '@fluentui/react-utilities';
import { useInput_unstable } from '@fluentui/react-input';
import { DismissRegular, SearchRegular } from '@fluentui/react-icons';
import type { ExtractSlotProps } from '@fluentui/react-utilities';
import type { SearchBoxSlots, SearchBoxProps, SearchBoxState } from './SearchBox.types';

/**
 * Create the state required to render SearchBox.
 *
 * The returned state can be modified with hooks such as useSearchBoxStyles_unstable,
 * before being passed to renderSearchBox_unstable.
 *
 * @param props - props from this instance of SearchBox
 * @param ref - reference to root HTMLElement of SearchBox
 */
export const useSearchBox_unstable = (props: SearchBoxProps, ref: React.Ref<HTMLInputElement>): SearchBoxState => {
  const { size = 'medium', disabled = false, root, contentBefore, dismiss, contentAfter, ...inputProps } = props;

  const searchBoxRootRef = React.useRef<HTMLDivElement>(null);
  const searchBoxRef = React.useRef<HTMLInputElement>(null);

  const [value, setValue] = useControllableState({
    state: props.value,
    defaultState: props.defaultValue,
    initialState: '',
  });

  // Tracks the focus of the component for the contentAfter and dismiss button
  const [focused, setFocused] = React.useState(false);

  const onFocus = useEventCallback(() => {
    setFocused(true);
  });

  const onBlur: React.FocusEventHandler<HTMLSpanElement> = useEventCallback(ev => {
    setFocused(!!searchBoxRootRef.current?.contains(ev.relatedTarget));
  });

  const rootProps = slot.resolveShorthand(root);

  const handleDismissClick = useEventCallback((event: React.MouseEvent<HTMLSpanElement>) => {
    if (isResolvedShorthand(dismiss)) {
      dismiss.onClick?.(event);
    }
    const newValue = '';
    setValue(newValue);
    props.onChange?.(event, { value: newValue });
  });

  const inputState = useInput_unstable(
    {
      type: 'search',
      disabled,
      size,
      value,
      root: slot.always<ExtractSlotProps<SearchBoxSlots['root']>>(
        {
          ...rootProps,
          ref: useMergedRefs(rootProps?.ref, searchBoxRootRef),
          onFocus: useEventCallback(mergeCallbacks(rootProps?.onFocus, onFocus)),
          onBlur: useEventCallback(mergeCallbacks(rootProps?.onBlur, onBlur)),
        },
        {
          elementType: 'span',
        },
      ),
      contentBefore: slot.optional(contentBefore, {
        renderByDefault: true,
        defaultProps: {
          children: <SearchRegular />,
        },
        elementType: 'span',
      }),
      contentAfter: slot.optional(contentAfter, {
        renderByDefault: true,
        elementType: 'span',
      }),
      ...inputProps,
      onChange: useEventCallback(ev => {
        const newValue = ev.target.value;
        props.onChange?.(ev, { value: newValue });
        setValue(newValue);
      }),
    },
    useMergedRefs(searchBoxRef, ref),
  );

  const state: SearchBoxState = {
    ...inputState,
    components: {
      ...inputState.components,
      dismiss: 'span',
    },
    dismiss: slot.optional(dismiss, {
      defaultProps: {
        children: <DismissRegular />,
        role: 'button',
        'aria-label': 'clear',
        tabIndex: -1,
      },
      renderByDefault: true,
      elementType: 'span',
    }),
    disabled,
    focused,
    size,
  };

  if (state.dismiss) {
    state.dismiss.onClick = handleDismissClick;
  }

  return state;
};
