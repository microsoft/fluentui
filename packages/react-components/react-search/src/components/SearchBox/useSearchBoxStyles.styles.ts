import { makeResetStyles, makeStyles, mergeClasses } from '@griffel/react';
import type { SearchBoxSlots, SearchBoxState } from './SearchBox.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';

export const searchBoxClassNames: SlotClassNames<SearchBoxSlots> = {
  root: 'fui-SearchBox',
  dismiss: 'fui-SearchBox__dismiss',
  contentAfter: 'fui-SearchBox__contentAfter',
};

/**
 * Styles for the root slot
 */
const useRootClassName = makeResetStyles({
  // removes the WebKit pseudoelement styling
  '::-webkit-search-decoration': {
    display: 'none',
  },
  '::-webkit-search-cancel-button': {
    display: 'none',
  },
});

const useContentClassName = makeResetStyles({
  boxSizing: 'border-box',
  color: tokens.colorNeutralForeground3, // "icon color" in design spec
  display: 'flex',
  // special case styling for icons (most common case) to ensure they're centered vertically
  // size: medium (default)
  '> svg': { fontSize: '20px' },
});

const useContentStyles = makeStyles({
  disabled: {
    color: tokens.colorNeutralForegroundDisabled,
  },
  // Ensure resizable icons show up with the proper font size
  small: {
    '> svg': { fontSize: '16px' },
  },
  medium: {
    // included in useContentClassName
  },
  large: {
    '> svg': { fontSize: '24px' },
  },
});

/**
 * Apply styling to the SearchBox slots based on the state
 */
export const useSearchBoxStyles_unstable = (state: SearchBoxState): SearchBoxState => {
  const { size } = state.root;
  const disabled = state.root.input!.disabled;

  const contentClassName = useContentClassName();
  const contentStyles = useContentStyles();

  state.root.input!.className = mergeClasses(useRootClassName(), state.root.className);

  if (state.dismiss) {
    state.dismiss.className = mergeClasses(
      searchBoxClassNames.dismiss,
      contentClassName,
      disabled && contentStyles.disabled,
      contentStyles[size],
      state.dismiss.className,
    );
  }

  return state;
};
