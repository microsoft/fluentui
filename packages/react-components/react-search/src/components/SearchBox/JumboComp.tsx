import * as React from 'react';
import { Input } from '@fluentui/react-input';
import { SearchBox } from './SearchBox';

export const JumboComp = (
  <>
    <Input />
    <SearchBox />
  </>
);
