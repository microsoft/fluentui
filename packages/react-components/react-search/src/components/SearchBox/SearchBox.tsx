import * as React from 'react';
import { useSearchBox_unstable } from './useSearchBox';
import { renderSearchBox_unstable } from './renderSearchBox';
import { useSearchBoxStyles_unstable } from './useSearchBoxStyles.styles';
import type { SearchBoxProps } from './SearchBox.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * SearchBox component - TODO: add more docs
 */
export const SearchBox: ForwardRefComponent<SearchBoxProps> = React.forwardRef((props, ref) => {
  const state = useSearchBox_unstable(props, ref);

  useSearchBoxStyles_unstable(state);
  return renderSearchBox_unstable(state);
});

SearchBox.displayName = 'SearchBox';
