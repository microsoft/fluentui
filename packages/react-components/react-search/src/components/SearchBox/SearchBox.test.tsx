import * as React from 'react';
import { render } from '@testing-library/react';
import { SearchBox } from './SearchBox';
import { isConformant } from '../../testing/isConformant';

describe('SearchBox', () => {
  isConformant({
    Component: SearchBox,
    displayName: 'SearchBox',
    primarySlot: 'input',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<SearchBox />);
    expect(result.container).toMatchSnapshot();
  });
});
