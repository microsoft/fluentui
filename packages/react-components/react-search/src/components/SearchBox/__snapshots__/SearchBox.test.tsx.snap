// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders a default state 1`] = `
<div>
  <span
    class="fui-Input fui-SearchBox"
  >
    <span
      class="fui-Input__contentBefore fui-SearchBox__contentBefore"
    >
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="1em"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.5 3a5.5 5.5 0 0 1 4.23 9.02l4.12 4.13a.5.5 0 0 1-.63.76l-.07-.06-4.13-4.12A5.5 5.5 0 1 1 8.5 3Zm0 1a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9Z"
          fill="currentColor"
        />
      </svg>
    </span>
    <input
      class="fui-Input__input fui-SearchBox__input"
      type="search"
      value=""
    />
    <span
      class="fui-Input__contentAfter fui-SearchBox__contentAfter"
    >
      <span
        aria-label="clear"
        class="fui-SearchBox__dismiss"
        role="button"
        tabindex="-1"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="currentColor"
          height="1em"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;
